x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -133,25 +195,75 @@ public void getPrivileges(Collection<String> applications, Collection<String> na
+            invalidateAll();
+        }
     }
 
-    private boolean hasWildcard(Collection<String> applications) {
-        return applications.stream().anyMatch(n -> n.endsWith("*"));
+    public void invalidate(Collection<String> updatedApplicationNames) {
+        logger.debug("Invalidating application privileges caches for: {}", updatedApplicationNames);
+        numInvalidation.incrementAndGet();
+        // Always completely invalidate application names cache due to wildcard
+        applicationNamesCache.invalidateAll();
+        final Set<String> uniqueNames = Set.copyOf(updatedApplicationNames);
+        uniqueNames.forEach(descriptorsCache::invalidate);
     }
 
-    private QueryBuilder getPrivilegeNameQuery(Collection<String> names) {
-        return QueryBuilders.termsQuery(ApplicationPrivilegeDescriptor.Fields.NAME.getPreferredName(), names);
+    public void invalidateAll() {
+        logger.debug("Invalidating all application privileges caches");
+        numInvalidation.incrementAndGet();
+        applicationNamesCache.invalidateAll();
+        descriptorsCache.invalidateAll();
+    }
+
+    /**
+     * Try resolve all privileges for given application names from the cache.
+     * It returns non-null result only when privileges of ALL applications are
+     * found in the cache, i.e. it returns null if any of application name is
+     * NOT found in the cache. Since the cached is keyed by concrete application
+     * name, this means any wildcard will result in null.
+     */
+    private Set<ApplicationPrivilegeDescriptor> cachedDescriptorsForApplicationNames(Set<String> applicationNames) {
+
+        final Set<ApplicationPrivilegeDescriptor> cachedDescriptors = new HashSet<>();
+        for (String applicationName: applicationNames) {
+            if (applicationName.endsWith("*")) {
+                return null;
+            } else {
+                final Set<ApplicationPrivilegeDescriptor> descriptors = descriptorsCache.get(applicationName);
+                if (descriptors == null) {
+                    return null;
+                } else {
+                    cachedDescriptors.addAll(descriptors);
+                }
+            }
+        }
+        return Collections.unmodifiableSet(cachedDescriptors);
+    }
+
+    /**
+     * Filter to get all privilege descriptors that have any of the given privilege names.
+     */
+    private Collection<ApplicationPrivilegeDescriptor> filterDescriptorsForPrivilegeNames(
+        Collection<ApplicationPrivilegeDescriptor> descriptors, Collection<String> privilegeNames) {
+        // empty set of names equals to retrieve everything
+        if (isEmpty(privilegeNames)) {
+            return descriptors;
+        }
+        return descriptors.stream().filter(d -> privilegeNames.contains(d.getName())).collect(Collectors.toUnmodifiableSet());
     }
 
     private QueryBuilder getApplicationNameQuery(Collection<String> applications) {
```

Comment:
Nit: I think the order of the methods in this file is a bit random.
This probably should be after `innerGetPrivileges` since it's an implementation detail of that method.

Reply:
Not sure If I understand correctly, but this method (L269) is already after `innerGetPrivileges` (L173). 

In terms of overall method ordering, I have seen two preferences: 1) order by visibility with public first and private last; 2) group by related functions with public methods followed by their supporting private methods. 
My impression is the ES code base loosely follows option 2. This file is example of it. Do we have a recommended style?

Reply:
Sorry, I was a bit loose with my terminology.
`getApplicationNameQuery` is _after_ `innerGetPrivileges`, but not _immediately after_.

I see 
- `innerGetPrivileges`
- `onSecurityIndexStateChange`
- `invalidate`
- `invalidateAll`
- `cachedDescriptorsForApplicationNames`
- `filterDescriptorsForPrivilegeNames`
- `getApplicationNameQuery`

I can't see any pattern to that order.

Reply:
Thanks for the elaboration. I tried to organise them by following option 2. Private methods are listed after the public methods that are the callers. The result is as follows:

```java
public void getPrivileges
private void innerGetPrivileges
private QueryBuilder getApplicationNameQuery
private ApplicationPrivilegeDescriptor buildPrivilege
private Set<ApplicationPrivilegeDescriptor> cachedDescriptorsForApplicationNames
private Collection<ApplicationPrivilegeDescriptor> filterDescriptorsForPrivilegeNames

public void putPrivileges
private void innerPutPrivilege

public void deletePrivileges
private <T> void clearCaches
private static Tuple<String, String> nameFromDocId
private static String toDocId

public void onSecurityIndexStateChange
public void invalidate
public void invalidateAll

private static boolean isEmpty
```

Does above ordering look reasonable to you?